WIP/Feedback wanted: Sink and Stream implementations

stkfd opened the pull request:
So as mentioned in #3 I've finally gotten to implementing Sink as well as Stream. This implementation works, however there are some open questions remaining:

* There's a bit of a disconnect between the requirements on the recv/send methods and the implementations for Sink/Stream. The implementations for Sink and Stream usually require a mutable reference, but that's not actually necessary for this channel because it internally clones the senders anyways. Maybe I'll try implementing the traits for &BroadcastChannel instead of BroadcastChannel?
* With the separate methods on Sink (poll_ready, start_send, poll_flush), I wonder what happens if a new clone of the channel is made anywhere between those calls. Need to write a test or two here I think.
* Related to the previous concern, I think it may be advisable to implement this in a way where the cloning doesn't happen on _every_ poll call, also for performance reasons.

Finally, an async port of crossbeam's channels just became available in the async-std project! It's MPMC, but _not_ a broadcast channel. Still may be worth seeing how that implementation works and if there are any good ideas to copy there.

stkfd committed:
support stream trait

stkfd committed:
Update example in readme

stkfd committed:
initial implementation for sink

leo60228 commented:
This looks good. I clone on every poll to reduce lock contention (since this library should, in theory, be able to handle networked scenarios).

stkfd committed:
Clean up Sink implementations, add Sink implementation for &BroadcastChannel

stkfd committed:
Merge branch 'master' of https://github.com/leo60228/broadcaster

stkfd committed:
fix imports for futures 0.3 stable

stkfd commented:
Forgot about this PR for a while there but I went back and cleaned up the sink implementation a bit and added a Sink implementation for &BroadcastChannel so that the `SinkExt` methods can be used with just an immutable reference to a channel. That implementation now does the real work. To make all the combinators (for example split) work there's still an implementation for the owned channel, but it just delegates to the borrowed version.
